src/Microsoft.AspNet.Mvc.Extensions/JsonResult.cs

```diff
@@ -81,80 +59,44 @@ public JsonResult(object value, IOutputFormatter formatter)
+                contentTypeHeader = DefaultContentType;
             }
             else
             {
-                objectResult.ContentTypes = ContentTypes;
+                if (contentTypeHeader.Encoding == null)
+                {
+                    // 1. Do not modify the user supplied content type
+                    // 2. Parse here to handle parameters apart from charset
+                    contentTypeHeader = MediaTypeHeaderValue.Parse(contentTypeHeader.ToString());
+                    contentTypeHeader.Encoding = Encoding.UTF8;
+                }
             }
 
-            var formatterContext = new OutputFormatterContext()
-            {
-                HttpContext = context.HttpContext,
-                DeclaredType = objectResult.DeclaredType,
-                Object = Value,
-            };
-
-            // JsonResult expects to always find a formatter, in contrast with ObjectResult, which might return
-            // a 406.
-            var formatter = SelectFormatter(objectResult, formatterContext);
-            Debug.Assert(formatter != null);
+            response.ContentType = contentTypeHeader.ToString();
 
             if (StatusCode != null)
             {
-                context.HttpContext.Response.StatusCode = StatusCode.Value;
+                response.StatusCode = StatusCode.Value;
             }
 
-            await formatter.WriteAsync(formatterContext);
-        }
-
-        private IOutputFormatter SelectFormatter(ObjectResult objectResult, OutputFormatterContext formatterContext)
-        {
-            if (Formatter == null)
+            using (var writer = new HttpResponseStreamWriter(response.Body, contentTypeHeader.Encoding))
             {
-                // If no formatter was provided, then run Conneg with the formatters configured in options.
-                var formatters = formatterContext
-                    .HttpContext
-                    .RequestServices
-                    .GetRequiredService<IOptions<MvcOptions>>()
-                    .Options
-                    .OutputFormatters
-                    .OfType<JsonOutputFormatter>()
-                    .ToArray();
-
-                var formatter = objectResult.SelectFormatter(formatterContext, formatters);
-                if (formatter == null)
+                using (var jsonWriter = new JsonTextWriter(writer))
                 {
-                    // If the available user-configured formatters can't write this type, then fall back to the
-                    // 'global' one.
-                    formatter = formatterContext
-                        .HttpContext
-                        .RequestServices
-                        .GetRequiredService<JsonOutputFormatter>();
-
-                    // Run SelectFormatter again to try to choose a content type that this formatter can do.
-                    objectResult.SelectFormatter(formatterContext, new[] { formatter });
+                    jsonWriter.CloseOutput = false;
```

Comment:
spoke to @kichalla, it is needed so we don't close the underlying response stream. MVC 5 has the same behavior.